File: Day46/Part3/shifting_realms.py
import heapq

def solve():
    # Hardcoded Input
    # 3 Realms, 4x4
    # 1 = Floor, 0 = Void
    
    # Realm A
    grid_a = [
        [1, 1, 0, 0],
        [0, 1, 1, 0],
        [0, 0, 1, 0],
        [0, 0, 1, 1]
    ]
    
    # Realm B (Bridge)
    grid_b = [
        [0, 0, 0, 0],
        [1, 1, 1, 1], # Bridge across row 1
        [1, 0, 0, 1],
        [1, 1, 1, 1]
    ]
    
    # Realm C (Goal)
    grid_c = [
        [0, 0, 0, 0],
        [0, 0, 0, 0],
        [0, 1, 1, 1],
        [0, 1, 0, 1]
    ]
    
    realms = [grid_a, grid_b, grid_c]
    N = len(grid_a)
    M = len(grid_a[0])
    
    # Dijkstra
    # State: (cost, r, c, realm_idx)
    pq = [(0, 0, 0, 0)]
    dist = {} # (r, c, realm) -> min_cost
    dist[(0, 0, 0)] = 0
    
    target = (N-1, M-1, 2) # Goal in Realm C
    
    min_cost = -1
    
    while pq:
        d, r, c, k = heapq.heappop(pq)
        
        if d > dist.get((r, c, k), float('inf')):
            continue
        
        if (r, c, k) == target:
            min_cost = d
            break
        
        # 1. Move within realm
        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = r + dr, c + dc
            if 0 <= nr < N and 0 <= nc < M:
                if realms[k][nr][nc] == 1:
                    new_cost = d + 1
                    if new_cost < dist.get((nr, nc, k), float('inf')):
                        dist[(nr, nc, k)] = new_cost
                        heapq.heappush(pq, (new_cost, nr, nc, k))
                        
        # 2. Shift Realm
        nk = (k + 1) % 3
        if realms[nk][r][c] == 1:
            new_cost = d + 3
            if new_cost < dist.get((r, c, nk), float('inf')):
                dist[(r, c, nk)] = new_cost
                heapq.heappush(pq, (new_cost, r, c, nk))
                
    print(f"Minimum Cost: {min_cost}")

if __name__ == "__main__":
    solve()

File: learnLeetCode/Day46/Part3/shifting_realms.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

const int INF = 1e9;

struct Node {
    int cost;
    int r;
    int c;
    int k; // Realm index
    
    bool operator>(const Node& other) const {
        return cost > other.cost;
    }
};

int main() {
    // Hardcoded Input
    int N = 4;
    int M = 4;
    
    vector<vector<vector<int>>> realms(3, vector<vector<int>>(N, vector<int>(M)));
    
    // Realm A
    realms[0] = {
        {1, 1, 0, 0},
        {0, 1, 1, 0},
        {0, 0, 1, 0},
        {0, 0, 1, 1}
    };
    
    // Realm B
    realms[1] = {
        {0, 0, 0, 0},
        {1, 1, 1, 1},
        {1, 0, 0, 1},
        {1, 1, 1, 1}
    };
    
    // Realm C
    realms[2] = {
        {0, 0, 0, 0},
        {0, 0, 0, 0},
        {0, 1, 1, 1},
        {0, 1, 0, 1}
    };
    
    // Dist array: [realm][r][c]
    vector<vector<vector<int>>> dist(3, vector<vector<int>>(N, vector<int>(M, INF)));
    priority_queue<Node, vector<Node>, greater<Node>> pq;
    
    dist[0][0][0] = 0;
    pq.push({0, 0, 0, 0});
    
    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};
    
    int minCost = -1;
    
    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        
        int d = current.cost;
        int r = current.r;
        int c = current.c;
        int k = current.k;
        
        if (d > dist[k][r][c]) continue;
        
        if (r == N-1 && c == M-1 && k == 2) {
            minCost = d;
            break;
        }
        
        // 1. Move within realm
        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            
            if (nr >= 0 && nr < N && nc >= 0 && nc < M) {
                if (realms[k][nr][nc] == 1) {
                    int newCost = d + 1;
                    if (newCost < dist[k][nr][nc]) {
                        dist[k][nr][nc] = newCost;
                        pq.push({newCost, nr, nc, k});
                    }
                }
            }
        }
        
        // 2. Shift Realm
        int nk = (k + 1) % 3;
        if (realms[nk][r][c] == 1) {
            int newCost = d + 3;
            if (newCost < dist[nk][r][c]) {
                dist[nk][r][c] = newCost;
                pq.push({newCost, r, c, nk});
            }
        }
    }
    
    cout << "Minimum Cost: " << minCost << endl;
    
    return 0;
}
